cmd/settlement_gemini.go

```diff
@@ -0,0 +1,318 @@
+
+	for _, payout := range *response {
+		if payout.Result == "Error" {
+			transactions["failed"] = append(transactions["failed"], originalTransactions[payout.TxRef])
+		} else {
+			if *payout.Status == "Pending" {
+				transactions["pending"] = append(transactions["pending"], originalTransactions[payout.TxRef])
+			} else {
+				transactions["completed"] = append(transactions["completed"], originalTransactions[payout.TxRef])
+			}
+		}
+	}
+	return transactions
+}
+
+// func convertTransactionListIntoMap(
+// 	transactions *[]gemini.PayoutPayload,
+// ) map[string]gemini.PayoutPayload {
+// 	transactionMap := make(map[string]gemini.PayoutPayload)
+// 	for _, payoutRequest := range *transactions {
+// 		transactionMap[payoutRequest.TxRef] = payoutRequest
+// 	}
+// 	return transactionMap
+// }
+
+// GeminiUploadSettlement marks the settlement file as complete
+func GeminiUploadSettlement(inPath string, signatureSwitch int, allTransactionsFile string, outPath string) error {
+	if outPath == "./gemini-settlement" {
+		// use a file with extension if none is passed
+		outPath = "./gemini-settlement-complete.json"
+	}
+
+	ctx := context.Background()
+	bulkPayoutFiles := strings.Split(inPath, ",")
+	geminiClient, err := gemini.New()
+	if err != nil {
+		return err
+	}
+
+	// transactionsForEyeshade := []settlement.Transaction{}
+	// failedTransactions := []settlement.Transaction{}
+	if allTransactionsFile == "" {
+		return errors.New("unable to upload without a transactions file to check against")
+	}
+	submittedTransactions := make(map[string][]settlement.Transaction)
+
+	bytes, err := ioutil.ReadFile(allTransactionsFile)
+	if err != nil {
+		return err
+	}
+
+	var settlementTransactions []settlement.AntifraudTransaction
+	err = json.Unmarshal(bytes, &settlementTransactions)
+	if err != nil {
+		return err
+	}
+	// create a map of the request transactions
+	transactionsMap := geminiMapTransactionsToID(settlementTransactions)
+
+	for _, bulkPayoutFile := range bulkPayoutFiles {
+		bytes, err := ioutil.ReadFile(bulkPayoutFile)
+		if err != nil {
+			return err
+		}
+
+		var geminiBulkPayoutRequestRequirements []gemini.PrivateRequestSequence
+		err = json.Unmarshal(bytes, &geminiBulkPayoutRequestRequirements)
+		if err != nil {
+			return err
+		}
+
+		for _, bulkPayoutRequestRequirements := range geminiBulkPayoutRequestRequirements {
+			// make sure payload is parsable
+			// upload the bulk payout
+			decodedSig, err := hex.DecodeString(bulkPayoutRequestRequirements.Signatures[signatureSwitch])
+			if err != nil {
+				return err
+			}
+			presigner := cryptography.NewPresigner(decodedSig)
+			bulkPayoutRequestRequirements.Base.Nonce = bulkPayoutRequestRequirements.Base.Nonce + int64(signatureSwitch)
+			serialized, err := json.Marshal(bulkPayoutRequestRequirements.Base)
+			if err != nil {
+				return err
+			}
+			payload := base64.StdEncoding.EncodeToString(serialized)
+			response, err := geminiClient.UploadBulkPayout(
```

Comment:
Are we doing any batching, or is this sending all of the payouts in one api call, or are all of the bulk payout files already chunked accordingly?

Reply:
bulk payout files should at this point be batched accordingly by the transform step. we can only upload 500 at a time

Reply:
:+1: